// chainlink/v0.5/dist/src/generated/CoordinatorInterfaceFactory.js
"use strict";
/* Generated by ts-generator ver. 0.0.8 */
/* tslint:disable */
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
class CoordinatorInterfaceFactory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.CoordinatorInterfaceFactory = CoordinatorInterfaceFactory;
const _abi = [
    {
        constant: false,
        inputs: [
            {
                name: "_requestId",
                type: "bytes32"
            },
            {
                name: "_aggregatorArgs",
                type: "bytes32"
            }
        ],
        name: "fulfillOracleRequest",
        outputs: [
            {
                name: "",
                type: "bool"
            }
        ],
        payable: false,
        stateMutability: "nonpayable",
        type: "function"
    },
    {
        constant: false,
        inputs: [
            {
                name: "_serviceAgreementData",
                type: "bytes"
            },
            {
                name: "_oracleSignaturesData",
                type: "bytes"
            }
        ],
        name: "initiateServiceAgreement",
        outputs: [
            {
                name: "",
                type: "bytes32"
            }
        ],
        payable: false,
        stateMutability: "nonpayable",
        type: "function"
    }
];
